packages/apollo-cache-control/README.md

```diff
@@ -113,3 +113,15 @@ app.use('/graphql', bodyParser.json(), graphqlExpress({
   },
 }));
 ```
+
+### Using custom header keys and/or values
+
+In certain situations you may want to specify the header key or value, for example, when using a CDN that requires a custom header for edge-caching. This can be achieved with the following configuration:
+
+```javascript
+cacheControl: {
+  calculateHttpHeaders: true,
+  headerKey: 'Edge-Control',
+  buildHeaderValue: (hint) => `!no-store, max-age=${hint.maxAge}`
```

Comment:
What if, rather than introducing `headerKey` and `buildHeaderValue`:

* `calculateHttpHeaders` was allowed to be `false`, `true`, or an `Object`.
* When `false` (i.e., `calculateHttpHeaders: false`), it would _NOT_ include the header.
* When defined as an `Object` (e.g., `calculateHttpHeaders: {}`), then the _keys_ would be cache-control headers and the corresponding _value_ for those keys would be functions to generate the value.  Your `buildHeaderValue` function would seem almost perfectly appropriate for this, though I'd suggest calling the argument `policy`, rather than `hint` since "hints" are what are applied to the schema and the "policy" is the computation of the overall determination based on those hints.  The type would, however, be a duplication of [the current `CacheHint` type](https://github.com/apollographql/apollo-server/blob/6f0dee7871d8cf06fa24b4242bb03f318a0697a8/packages/apollo-cache-control/src/index.ts#L16-L19):

  ```typescript
  calculateHttpHeaders: {
    "Edge-Control": (policy: CachePolicy) => `!no-store, max-age=${hint.maxAge}`,
  },
  ```
* When it was `true`, it would continue to calculate `cache-control` exactly as it did before.  It would be syntactic sugar for:

  ```typescript
  calculateHttpHeaders: {
    "Cache-Control": (policy: CachePolicy) =>
                       `max-age=${hint.maxAge}, ${policy.scope.toLowerCase()}`,
  },
  ```
* I think there's (maybe?) even an option here where the values on the `Object` could be `Boolean` values to enable the default header.  That would enable someone to have a `Cache-control` header that uses the default Apollo Server behavior _in addition to_ a `Edge-Control` that did something custom:

  ```typescript
  calculateHttpHeaders: {
    "Edge-Control": (policy: CachePolicy) =>
                       `!no-store, max-age=${hint.maxAge}`,
    "Cache-Control": true,
  },
  ```


I think this also allows us to keep all HTTP-transport specific preferences within an HTTP-specific property and off the top-level options (since this package is theoretically intended for general cache control use, not just over HTTP).

Reply:
Thanks for the input! This certainly seems like a much more flexible approach. I'll put some time aside for this at the weekend :)

Reply:
I'm not sure, if this is the right PR to piggyback on or I should create a new Issue/PR for this topic.

This [RFC](https://tools.ietf.org/html/rfc5861) introduces the idea of the HTTP Cache-Control Extensions for Stale Content `stale-while-revalidate` and `stale-if-error`.

In order to balance immediacy and freshness when serving our content, we would like to add the directive `stale-while-revalidate=<seconds>` to the cache control header. You should probably be able to set this on a field level. It could make use of most of the logic, that is already in place for `maxAge` (e.g.: finding the lowest value).

For the second directive `stale-if-error` a different approach might be more suitable.
We are currently adding it on a request level. Whenever the response does not include an error object, we add the `stale-if-error` directive to the `cache-control` header.


It would be nice if we could add this to the apollo-cache-control package.

Reply:
Hey @dmateiu, you raise a really good point with this! I'm certain that I will need this functionality also. I think that given the length of time that this PR has been open, we should get it closed as-is and open a new PR for the `stale-if-error` work where we can properly document an API. What are your thoughts?